### PR TITLE
Full Python 3 compatibility

### LINKED ISSUE #58: PyLTI needs does not compile with pytest >= 2.8.0
PyLTI will need to be updated at some point as the tests fail with new versions of pytest.

### PR DESCRIPTION
**Pull request by GuillaumeDerval**
I spent some hours to make PyLTI pass the tests again with Python 3 (tested with 2.7, 3.4 and 3.5, I'm not sure it will pass with Python 3.1, 3.2 and 3.3 since they don't support the u prefix from what I remember)

I dropped oauth, since oauth2 is a fork, but oauth2 itself has some problems, which I had to fix manually (see classes SignatureMethod_HMAC_SHA1_Unicode, SignatureMethod_PLAINTEXT_Unicode, Request_Fix_Duplicate) (joestump/python-oauth2#207 joestump/python-oauth2#197).

It fixes #58 as a bonus too.

100% of the test passes with both python 2.7, 3.4 and 3.5, and the lib was tested on edX with Python 3.5, it works flawlessly.

### CODE REVIEW

**Comment by pdpinch** [comment]
@amir-qayyum-khan (or maybe @giocalitri if you have some downtime) can you look at this?

**Comment by giocalitri** [comment]
I will take a look

File: pylti/common.py [hunk 1]
```diff
@@ -4,12 +4,19 @@
...
 import oauth2
-import oauth.oauth as oauth
 from xml.etree import ElementTree as etree
 
+from oauth2 import STRING_TYPES
+try:  # Python 2
+    from urlparse import urlparse
+    from urllib import urlencode
+except ImportError:  # Python 3
+    from urllib.parse import urlparse, urlencode
```
Comment (giocalitri) [inline pylti/common.py]: I think we should use the [six library](https://pythonhosted.org/six/) for these imports

File: pylti/common.py [hunk 2]
```diff
@@ -49,18 +56,12 @@
...
 
-class LTIOAuthDataStore(oauth.OAuthDataStore):
-    # pylint: disable=abstract-method
-    """
-    Largely taken from reference implementation
-    for app engine at https://code.google.com/p/ims-dev/
-    """
-
-    def __init__(self, consumers):
+class LTIOAuthServer(oauth2.Server):
```
Comment (giocalitri) [inline pylti/common.py]: please modify the docstring accordingly without removing it

File: pylti/common.py [hunk 3]
```diff
@@ -301,11 +291,15 @@ def verify_request_common(consumers, url, method, headers, params):
...
-    except oauth.OAuthError:
+        oauth_consumer_key = oauth_request.get_parameter('oauth_consumer_key')
+        consumer = oauth_server.lookup_consumer(oauth_consumer_key)
+        if not consumer:
+            raise oauth2.Error('Invalid consumer.')
+        oauth_server.verify_request(oauth_request, consumer, None)
+    except oauth2.Error as e:
         # Rethrow our own for nice error handling (don't print
         # error message as it will contain the key
+        print(e)
```
Comment (giocalitri) [inline pylti/common.py]: what's the purpose of this `print`?

File: pylti/common.py [hunk 4]
```diff
@@ -350,7 +344,91 @@ def generate_request_xml(message_identifier_id, operation,
...
+
+class SignatureMethod_HMAC_SHA1_Unicode(oauth2.SignatureMethod_HMAC_SHA1):
+    """
+    Temporary workaround for
+    https://github.com/joestump/python-oauth2/issues/207
+    """
+
+    def check(self, request, consumer, token, signature):
+        """Returns whether the given signature is the correct signature for
+        the given consumer and token signing the given request."""
```
Comment (giocalitri) [inline pylti/common.py]: nit, please cane docstring to this:
```
"""
Returns whether the given signature is the correct signature for
the given consumer and token signing the given request.
"""

File: pylti/common.py [hunk 5]
```diff
@@ -350,7 +344,91 @@ def generate_request_xml(message_identifier_id, operation,
...
+
+class SignatureMethod_PLAINTEXT_Unicode(oauth2.SignatureMethod_PLAINTEXT):
+    """
+    Temporary workaround for
+    https://github.com/joestump/python-oauth2/issues/207
+    """
+
+    def check(self, request, consumer, token, signature):
+        """Returns whether the given signature is the correct signature for
+        the given consumer and token signing the given request."""
```
Comment (giocalitri) [inline pylti/common.py]: please fix docstring

File: pylti/common.py [hunk 6]
```diff
@@ -350,7 +344,91 @@ def generate_request_xml(message_identifier_id, operation,
...
+        """Return a string that contains the parameters that must be signed."""
+        items = []
+        for key, value in self.items():
+            if key == 'oauth_signature':
+                continue
+            # 1.0a/9.1.1 states that kvp must be sorted by key, then by value,
+            # so we unpack sequence values into multiple items for sorting.
+            if isinstance(value, STRING_TYPES):
+                items.append((oauth2.to_utf8_if_string(key),
+                              oauth2.to_utf8(value)))
```
Comment (giocalitri) [inline pylti/common.py]: nit: please write this line like
```
items.append(
    (oauth2.to_utf8_if_string(key), oauth2.to_utf8(value))
)
```

File: pylti/common.py [hunk 7]
```diff
@@ -350,7 +344,91 @@ def generate_request_xml(message_identifier_id, operation,
...
+            else:
+                try:
+                    value = list(value)
+                except TypeError as e:
+                    assert 'is not iterable' in str(e)
```
Comment (giocalitri) [inline pylti/common.py]: why do you need this `assert` ?

File: pylti/common.py [hunk 8]
```diff
@@ -350,7 +344,91 @@ def generate_request_xml(message_identifier_id, operation,
...
+                    items.append((oauth2.to_utf8_if_string(key),
+                                  oauth2.to_utf8_if_string(value)))
```
Comment (giocalitri) [inline pylti/common.py]: please indent as suggested above

File: pylti/common.py [hunk 9]
```diff
@@ -350,7 +344,91 @@ def generate_request_xml(message_identifier_id, operation,
...
+                else:
+                    items.extend((oauth2.to_utf8_if_string(key),
+                                  oauth2.to_utf8_if_string(item))
+                                 for item in value)
```
Comment (giocalitri) [inline pylti/common.py]: please indent in the following way:
```
items.extend(
    (oauth2.to_utf8_if_string(key), oauth2.to_utf8_if_string(item)) for item in value
)
```

File: pylti/tests/test_common.py [hunk 10]
```diff
@@ -7,12 +7,16 @@
...
 import httpretty
 import oauthlib.oauth1
-from urlparse import urlparse, parse_qs
-import urllib
+
+try:  # Python 2
+    from urlparse import urlparse, parse_qs
+    from urllib import urlencode
+except ImportError:  # Python 3
+    from urllib.parse import urlparse, parse_qs, urlencode
```
Comment (giocalitri) [inline pylti/tests/test_common.py]: please use six

File: pylti/tests/test_flask.py [hunk 11]
```diff
@@ -4,12 +4,16 @@
...
-import urllib
 
 import httpretty
 import mock
 import oauthlib.oauth1
 
+try:  # Python 2
+    from urllib import urlencode
+except ImportError:  # Python 3
+    from urllib.parse import urlencode
```
Comment (giocalitri) [inline pylti/tests/test_flask.py]: please use six

File: pylti/common.py (hunk 7 above)
Comment (GuillaumeDerval) [inline pylti/common.py]: This code is a slight adaptation of the method with the same name in python-oauth2. It is in fact the function patched with https://github.com/joestump/python-oauth2/pull/197.

The assert is part of the original code.

I forgot to add the specific copyright notice; I will do it with the remaining corrections :-)

File: pylti/common.py (hunk 9 above)
Comment (GuillaumeDerval) [inline pylti/common.py]: Line becomes too long.

```
items.extend(
    (oauth2.to_utf8_if_string(key),
     oauth2.to_utf8_if_string(item))
    for item in value
)
```

Is this ok?

**Comment by giocalitri** [comment]
@GuillaumeDerval thanks a lot for your contribution, we really appreciated.

I had some minor comments that actually made me think I should probably fix some other minor things in this project eventually.

Thanks again.

**Comment by GuillaumeDerval** [comment]
I applied most of your comments, and updated the README :-)

Would you like that I squash the commits together?

**Comment by giocalitri** [comment]
Thanks for the changes. 
Yes, please squash.
We will need to make some tests,so the actual merge can take some time.

File: pylti/common.py (hunk 7 above)
Comment (giocalitri) [inline pylti/common.py]: Ok, but I still think that line is unnecessary here.

**Comment by pdpinch** [comment]
@giocalitri I took a look at MIT-licensed code ([here](https://github.com/mitodl/pylti/pull/61/files#diff-3c41bec7c8b1cdd5800a9e5e4c3d89a6R358) and [here](https://github.com/mitodl/pylti/pull/61/files#diff-3c41bec7c8b1cdd5800a9e5e4c3d89a6R377)) and it's compatible with our BSD license. It's fine to merge from a licensing perspective. 

I still need to find a good test case.

**Comment by GuillaumeDerval** [comment]
Any news on this? It would be nice a py3 compatible version on pip :-)

**Comment by pdpinch** [comment]
@giocalitri can we merge & release this?

**Comment by perllaghu** [comment]
I was about to raise an issue to address this very issue - do we have any movement on this?

**Comment by pdpinch** [comment]
It's passed code review, but we haven't had time to do a functional test. Would you mind giving this branch a try and letting us know if it works for you?

**Comment by perllaghu** [comment]
It installs just fine in a virtualenv, however it's not installing into a docker container:

Create a Dockerfile in an otherwise empty directory
```Dockerfile
FROM ubuntu:16.04

RUN apt-get update && apt-get install -y \
  apache2 \
  apache2-utils \
  libssl-dev \
  apache2-dev \
  git \
  autoconf \
  python3.5 \
  libapache2-mod-wsgi-py3 \
  python3-pip \
  postgresql \
  postgresql-contrib \
  pwgen

# Apache logs
RUN ln -sf /proc/self/fd/1 /var/log/apache2/access.log && \
ln -sf /proc/self/fd/1 /var/log/apache2/error.log

# Load Python modules from requirements.txt
RUN pip3 install git+https://github.com/GuillaumeDerval/pylti@py3-restore#egg=PyLTI
```

then run
`docker build  .`

produces an error:
```
Collecting PyLTI from git+https://github.com/GuillaumeDerval/pylti@py3-restore#egg=PyLTI
  Cloning https://github.com/GuillaumeDerval/pylti (to py3-restore) to /tmp/pip-build-homdxv96/PyLTI
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-homdxv96/PyLTI/setup.py", line 121, in <module>
        README = open('README.rst').read()
      File "/usr/lib/python3.5/encodings/ascii.py", line 26, in decode
        return codecs.ascii_decode(input, self.errors)[0]
    UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 2423: ordinal not in range(128)
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-homdxv96/PyLTI/
You are using pip version 8.1.1, however version 9.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
The command '/bin/sh -c pip3 install git+https://github.com/GuillaumeDerval/pylti@py3-restore#egg=PyLTI' returned a non-zero code: 1
```

**Comment by perllaghu** [comment]
We needed to add

    RUN locale-gen en_GB.UTF-8
    ENV LC_ALL en_GB.UTF-8

to the Dockerfile, and then we could add `git+https://github.com/GuillaumeDerval/pylti@py3-restore#egg=PyLTI` to the `requirements.txt` file for the Django app

The app installs, and _django-lti-provider_ installs fine.

I've not yet tested any LTI functionality though.

**Comment by nikolas** [comment]
I'm trying to use this library in Python 3 - can someone merge this and make a new release?

**Comment by nikolas** [comment]
FYI - until this is merged and released, I've forked pylti with these changes and made a new release here: https://github.com/ccnmtl/pylti/releases/tag/0.5.0

**Comment by pdpinch** [comment]
I'm sorry that this PR has been neglected @nikolas. Were you able to use this branch as is, or did you have to make further changes?

**Comment by nikolas** [comment]
No problem, it happens. Yeah, this branch is working as is for me, if that helps.

**Comment by giocalitri** [comment]
I am going to merge this.
Thanks @GuillaumeDerval for the contribution and everybody for the patience.

**Pull request merged**